File: apps/desktop/lib/screens/device_pairing_screen.dart
import 'dart:async';
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:provider/provider.dart';
import 'package:qr_flutter/qr_flutter.dart';
import '../design/nano_tokens.dart';
import '../providers/app_provider.dart';
import '../services/server_service.dart';
import '../widgets/nano/nano_card.dart';
import '../widgets/nano/nano_primitives.dart';

/// Generate a pairing QR + 6-digit code so another device can join a server.
///
/// Requires an account-authenticated connection (the `/devices/token` endpoint
/// needs a JWT; device-token-only connections can't generate codes).
class DevicePairingScreen extends StatefulWidget {
  final String serverId;
  const DevicePairingScreen({super.key, required this.serverId});

  @override
  State<DevicePairingScreen> createState() => _DevicePairingScreenState();
}

class _DevicePairingScreenState extends State<DevicePairingScreen> {
  Future<DeviceTokenResult?>? _future;

  /// Client-side fallback TTL, used only when the server's response carries no
  /// real `expiresAt` (mirrors device_service.go `pairingCodeTTL`).
  static const _ttl = Duration(minutes: 15);

  /// Permission level the freshly-generated device token grants, taken from the
  /// /devices/token response's `permissionLevel`. Defaults to read-only (the
  /// server's least-privilege default) until the request resolves.
  int _grantedLevel = 0;

  Timer? _countdown;
  int _remaining = 0; // seconds until the current code expires

  @override
  void initState() {
    super.initState();
    _generate();
  }

  @override
  void dispose() {
    _countdown?.cancel();
    super.dispose();
  }

  void _generate() {
    final provider = context.read<AppProvider>();
    final svc = provider.serviceForServer(widget.serverId);
    final name = provider.getServerName(widget.serverId);
    final future =
        svc == null ? Future<DeviceTokenResult?>.value(null) : svc.generateDeviceToken(serverName: name);
    setState(() {
      _future = future;
      _grantedLevel = 0;
    });
    // Start the client-estimated countdown immediately; once the response
    // resolves, re-seed it from the real expiry / permission level.
    _startCountdown();
    future.then((r) {
      if (!mounted || r == null) return;
      setState(() => _grantedLevel = r.permissionLevel);
      if (r.expiresAt != null) _startCountdown(until: r.expiresAt);
    });
  }

  /// (Re)start the expiry countdown. When [until] is supplied the remaining
  /// time is derived from the server's real expiry; otherwise it falls back to
  /// the client-side [_ttl] estimate.
  void _startCountdown({DateTime? until}) {
    _countdown?.cancel();
    _remaining = until != null
        ? until.difference(DateTime.now()).inSeconds.clamp(0, _ttl.inSeconds * 4)
        : _ttl.inSeconds;
    _countdown = Timer.periodic(const Duration(seconds: 1), (_) {
      if (!mounted) return;
      setState(() {
        if (_remaining > 0) _remaining--;
      });
      if (_remaining == 0) _countdown?.cancel();
    });
  }

  String _fmtRemaining() {
    final m = _remaining ~/ 60;
    final s = (_remaining % 60).toString().padLeft(2, '0');
    return '$m:$s';
  }

  @override
  Widget build(BuildContext context) {
    final t = context.nano;
    final serverName = context.read<AppProvider>().getServerName(widget.serverId);
    return Scaffold(
      backgroundColor: t.bg,
      appBar: AppBar(
        backgroundColor: t.bg,
        elevation: 0,
        leading: IconButton(
          icon: Icon(Icons.arrow_back_rounded, color: t.fg),
          onPressed: () => Navigator.pop(context),
        ),
        title: Text('pairing.title'.tr(),
            style: TextStyle(
                fontSize: 18, fontWeight: FontWeight.w600, color: t.fg)),
      ),
      body: SafeArea(
        top: false,
        child: FutureBuilder<DeviceTokenResult?>(
          future: _future,
          builder: (context, snap) {
            if (snap.connectionState == ConnectionState.waiting) {
              return const Center(child: CircularProgressIndicator());
            }
            final r = snap.data;
            if (r == null) {
              return _error(t);
            }
            return _content(t, r, serverName);
          },
        ),
      ),
    );
  }

  Widget _content(NanoTokens t, DeviceTokenResult r, String serverName) {
    return ListView(
      padding: const EdgeInsets.fromLTRB(20, 8, 20, 32),
      children: [
        Text('pairing.intro'.tr(namedArgs: {'name': serverName}),
            style: TextStyle(fontSize: 14, color: t.fg3, height: 1.5)),
        const SizedBox(height: 20),
        Center(
          child: Container(
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.circular(20),
            ),
            child: QrImageView(
              data: r.qrData,
              version: QrVersions.auto,
              size: 220,
              backgroundColor: Colors.white,
              eyeStyle: const QrEyeStyle(
                eyeShape: QrEyeShape.square,
                color: Color(0xFF0A0A0A),
              ),
              dataModuleStyle: const QrDataModuleStyle(
                dataModuleShape: QrDataModuleShape.square,
                color: Color(0xFF0A0A0A),
              ),
            ),
          ),
        ),
        const SizedBox(height: 24),
        NanoSectionLabel('pairing.pairingCode'.tr(), grouped: true),
        NanoCard(
          child: NanoListRow(
            divider: false,
            trailing: IconButton(
              icon: Icon(Icons.copy_rounded, color: t.accent, size: 20),
              onPressed: () {
                Clipboard.setData(ClipboardData(text: r.pairingCode));
                ScaffoldMessenger.of(context).showSnackBar(
                    SnackBar(content: Text('pairing.codeCopied'.tr())));
              },
            ),
            child: Text(_fmtCode(r.pairingCode),
                style: TextStyle(
                    fontSize: 26,
                    fontWeight: FontWeight.w700,
                    letterSpacing: 6,
                    color: t.fg,
                    fontFamilyFallback: kMonoFallback)),
          ),
        ),
        const SizedBox(height: 12),
        _metaRow(t),
        const SizedBox(height: 16),
        Container(
          padding: const EdgeInsets.all(14),
          decoration: BoxDecoration(
            color: t.warn.withValues(alpha: 0.1),
            borderRadius: BorderRadius.circular(12),
          ),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Icon(Icons.lock_outline_rounded, color: t.warn, size: 18),
              const SizedBox(width: 10),
              Expanded(
                child: Text(
                    'pairing.securityNote'.tr(),
                    style: TextStyle(fontSize: 12.5, color: t.fg2, height: 1.45)),
              ),
            ],
          ),
        ),
        const SizedBox(height: 20),
        SizedBox(
          height: 48,
          child: OutlinedButton.icon(
            onPressed: _generate,
            icon: Icon(Icons.refresh_rounded, size: 18, color: t.accent),
            label: Text('pairing.regenerate'.tr(), style: TextStyle(color: t.accent)),
            style: OutlinedButton.styleFrom(
              side: BorderSide(color: t.sep),
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(t.buttonRadius),
              ),
            ),
          ),
        ),
      ],
    );
  }

  /// Expiry countdown for the 15-minute code TTL + the permission level the
  /// paired device will be granted.
  Widget _metaRow(NanoTokens t) {
    final expired = _remaining == 0;
    final low = _remaining > 0 && _remaining < 60;
    final expColor = expired
        ? t.crit
        : low
            ? t.warn
            : t.fg2;
    return Row(
      children: [
        Icon(Icons.timer_outlined, size: 16, color: expColor),
        const SizedBox(width: 6),
        Text(
          expired
              ? 'pairing.expired'.tr()
              : 'pairing.expiresIn'
                  .tr(namedArgs: {'time': _fmtRemaining()}),
          style: TextStyle(
              fontSize: 13,
              color: expColor,
              fontFamilyFallback: kMonoFallback),
        ),
        const Spacer(),
        Text('pairing.grants'.tr(),
            style: TextStyle(fontSize: 12.5, color: t.fg3)),
        const SizedBox(width: 8),
        NanoPermPill(level: _grantedLevel),
      ],
    );
  }

  Widget _error(NanoTokens t) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.all(32),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(Icons.qr_code_2_rounded, size: 36, color: t.fg4),
            const SizedBox(height: 12),
            Text('pairing.errorTitle'.tr(),
                style: TextStyle(
                    fontSize: 16, fontWeight: FontWeight.w600, color: t.fg)),
            const SizedBox(height: 6),
            Text('pairing.errorDesc'.tr(),
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 13, color: t.fg3, height: 1.45)),
            const SizedBox(height: 16),
            TextButton.icon(
              onPressed: _generate,
              icon: Icon(Icons.refresh_rounded, size: 16, color: t.accent),
              label: Text('common.retry'.tr(), style: TextStyle(color: t.accent)),
            ),
          ],
        ),
      ),
    );
  }

  String _fmtCode(String code) {
    if (code.length == 6) return '${code.substring(0, 3)} ${code.substring(3)}';
    return code;
  }
}

File: apps/ios/NanoOps/Screens/DevicePairingScreen.swift
import SwiftUI
import UIKit

/// Generates a short-lived device token and presents its QR payload + six-digit
/// pairing code. Requires an account-authenticated server connection.
struct DevicePairingScreen: View {
    let serverId: String

    @EnvironmentObject private var store: AppStore
    @Environment(\.nano) private var t
    @State private var result: DeviceTokenResult?
    @State private var loading = true
    @State private var remaining = 0
    @State private var countdownTask: Task<Void, Never>?

    private let fallbackTTL = 15 * 60

    var body: some View {
        Group {
            if loading {
                ProgressView().controlSize(.large)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else if let result {
                content(result)
            } else {
                errorView
            }
        }
        .background(t.bg.ignoresSafeArea())
        .navigationTitle(tr("pairing.title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { if result == nil { await generate() } }
        .onDisappear { countdownTask?.cancel() }
    }

    private func content(_ result: DeviceTokenResult) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text(tr("pairing.intro", ["name": store.serverName(serverId)]))
                    .font(.system(size: 14))
                    .foregroundColor(t.fg3)
                    .lineSpacing(5)

                QRCodeView(value: result.qrData)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 20)

                NanoSectionLabel(tr("pairing.pairingCode"), grouped: true)
                NanoCard {
                    NanoListRow(divider: false) {
                        Text(formatCode(result.pairingCode))
                            .font(NanoFont.mono(26, weight: .bold))
                            .tracking(6)
                            .foregroundColor(t.fg)
                    } trailing: {
                        Button {
                            UIPasteboard.general.string = result.pairingCode
                        } label: {
                            Image(systemName: "doc.on.doc")
                                .font(.system(size: 18))
                                .foregroundColor(t.accent)
                                .frame(width: 36, height: 36)
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel(tr("common.copy"))
                    }
                }

                metadataRow(level: result.permissionLevel)
                    .padding(.top, 12)

                HStack(alignment: .top, spacing: 10) {
                    Image(systemName: "lock")
                        .font(.system(size: 16))
                        .foregroundColor(t.warn)
                    Text(tr("pairing.securityNote"))
                        .font(.system(size: 12.5))
                        .foregroundColor(t.fg2)
                        .lineSpacing(4)
                }
                .padding(14)
                .background(t.warn.opacity(0.1))
                .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                .padding(.top, 16)

                NanoButton(tr("pairing.regenerate"), icon: "arrow.clockwise",
                           variant: .outlined, fullWidth: true, loading: loading) {
                    Task { await generate() }
                }
                .padding(.top, 20)
            }
            .padding(EdgeInsets(top: 8, leading: 20, bottom: 32, trailing: 20))
        }
    }

    private func metadataRow(level: Int) -> some View {
        let expired = remaining == 0
        let low = remaining > 0 && remaining < 60
        let color = expired ? t.crit : (low ? t.warn : t.fg2)
        return HStack(spacing: 6) {
            Image(systemName: "timer").font(.system(size: 15)).foregroundColor(color)
            Text(expired ? tr("pairing.expired") : tr("pairing.expiresIn", ["time": formatRemaining()]))
                .font(NanoFont.mono(13))
                .foregroundColor(color)
            Spacer()
            Text(tr("pairing.grants")).font(.system(size: 12.5)).foregroundColor(t.fg3)
            NanoPermPill(level: level)
        }
    }

    private var errorView: some View {
        VStack(spacing: 12) {
            Image(systemName: "qrcode")
                .font(.system(size: 36))
                .foregroundColor(t.fg4)
            Text(tr("pairing.errorTitle"))
                .font(.system(size: 16, weight: .semibold))
                .foregroundColor(t.fg)
            Text(tr("pairing.errorDesc"))
                .font(.system(size: 13))
                .foregroundColor(t.fg3)
                .multilineTextAlignment(.center)
                .lineSpacing(4)
            NanoButton(tr("common.retry"), icon: "arrow.clockwise", variant: .text) {
                Task { await generate() }
            }
        }
        .padding(32)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    @MainActor
    private func generate() async {
        countdownTask?.cancel()
        loading = true
        result = nil
        guard let service = store.serviceForServer(serverId) else {
            loading = false
            return
        }
        let generated = await service.generateDeviceToken(serverName: store.serverName(serverId))
        result = generated
        loading = false
        guard let generated else { return }
        let seconds: Int
        if let expiresAt = generated.expiresAt {
            seconds = min(max(Int(expiresAt.timeIntervalSinceNow), 0), fallbackTTL * 4)
        } else {
            seconds = fallbackTTL
        }
        startCountdown(seconds)
    }

    private func startCountdown(_ seconds: Int) {
        remaining = seconds
        countdownTask = Task {
            while !Task.isCancelled && remaining > 0 {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard !Task.isCancelled else { return }
                remaining -= 1
            }
        }
    }

    private func formatRemaining() -> String {
        String(format: "%d:%02d", remaining / 60, remaining % 60)
    }

    private func formatCode(_ code: String) -> String {
        guard code.count == 6 else { return code }
        let split = code.index(code.startIndex, offsetBy: 3)
        return "\(code[..<split]) \(code[split...])"
    }
}
